documents/tree.go

```diff
@@ -29,13 +43,32 @@ func NewLeafProperty(literal string, compact []byte) proofs.Property {
 	return proofs.NewProperty(literal, compact...)
 }
 
-// GenerateNewSalts generates salts for new Document
-func GenerateNewSalts(document proto.Message, prefix string, compactPrefix []byte) (*proofs.Salts, error) {
-	docSalts := new(proofs.Salts)
-	t := NewDefaultTreeWithPrefix(docSalts, prefix, compactPrefix)
-	err := t.AddLeavesFromDocument(document)
-	if err != nil {
-		return nil, err
-	}
-	return docSalts, nil
+// DocumentSaltsFunc returns a function that fetches and sets salts on the CoreDoc. The boolean mutable can be used to define if the salts function should error if a new field is encountered or not.
+func DocumentSaltsFunc(coredoc *coredocumentpb.CoreDocument, mutable bool) (func(compact []byte) ([]byte, error), error) {
```

Comment:
Let's pass only the pointer to the salts instead of passing the full coredocument. That will make the functions more generic and single source to generate trees of various documents